rest_api/limsdb/queries/data_models.py

```diff
@@ -319,3 +320,69 @@ def to_json(self):
             'nb_reads': self.udfs['Read'],
             'nb_cycles': self.udfs['Cycle']
         }
+
+
+class Library:
+    def __init__(self):
+        self.id = None
+        self.type = None
+        self.preps = defaultdict(LibraryPrep)
+        self.projects = set()
+
+    def to_json(self):
+        most_recent_prep = sorted(self.preps.values(), key=lambda p: p.date_run, reverse=True)[0]
+        data = {
+            'id': self.id,
+            'times_prepared': len(self.preps),
+            'last_prepared': format_date(most_recent_prep.date_run),
+            'qc': {},
+            'project_ids': sorted(self.projects),
+            'type': status_cfg.protocol_names.get(self.type, 'unknown'),
+            'nsamples': len(most_recent_prep.placements)
+        }
+
+        passing_samples = 0
+        for k, v in most_recent_prep.placements.items():
+            data['qc']['%s:%s' % k] = v.to_json()
+            if v.qc_flag == 'PASSED':
+                passing_samples += 1
+
+        data['pc_qc_flag_pass'] = (passing_samples / len(most_recent_prep.placements)) * 100
+        return data
+
+
+class LibraryPrep:
+    def __init__(self):
+        self.id = None
+        self.date_run = None
+        self.placements = defaultdict(LibrarySample)
+
+
+class LibrarySample:
+    state_map = {
+        0: 'UNKNOWN',
+        1: 'PASSED',
+        2: 'FAILED'
+    }
+
+    def __init__(self):
+        self.name = None
+        self.udf = {}
+        self.states = {}
+
+    @cached_property
+    def qc_flag(self):
+        """Duplicates functionality in genologics_sql.tables.Artifact.qc_flag"""
```

Comment:
I've just checked the implementation in genologics_sql, and it is different (although it probably results in similar results). Are these two functions intended to be used interchangeably? Why was it necessary to duplicate the functionality?

Reply:
I can't use the one in genologics_sql in the middle of a database query. I might be able to use it if we pass SQLAlchemy objects to `factories`